### ISSUE TITLE
Merge similar colors in Toonz level. 

### ISSUE THREAD

**Issue by F-Burning** [opentoonz/opentoonz#4369]
I you paste a color from another level with a drawing, and repeat this along some frames, it creates new colors instead of paste it in one style. I think the skill to merge all similar levels in one, or paste drawings with a color into styles, would improve workflow.

**Comment by shun-iwasawa** [comment]
Currently when pasting frame or pasting raster selection from one toonz raster level to another, the used style in the pasted image will be merged to the destination level's existing style only when 1) the styles are identical (i.e. with the same names and colors) and 2) the style indices are the same.

**Comment by F-Burning** [comment]
> Currently when pasting frame or pasting raster selection from one toonz raster level to another, the used style in the pasted image will be merged to the destination level's existing style only when 1) the styles are identical (i.e. with the same names and colors) and 2) the style indices are the same.

True, interesting to know, thanks.

Anyway, sometimes you forget follow this steps because are very specific, and styles repeat on palette, so merge them would be a solution.

**Comment by F-Burning** [comment]
By the way, it's technically possible paste selections in different styles? I asked this time ago, but It not was clear for me if this is not possible to Opentoonz, or is a decision.

**Comment by RodneyBaker** [comment]
@F-Burning You may be interested in @Wolfinabowl's utility that can convert palettes, merge colors etc.
He has the uility (and others) on twitch.io and while it has a place for donations he's fine with people entering $0 to grab the program.  Off hand I don't recall the exact merging capabilities but it's a great utility to have standing by.  

Here's what he's written about the utility:

> Display, modify, create, and convert between a variety of palette formats.
> 
> Supported formats:
> 
> tpl - OpenToonz, Tahoma
> gpl - Krita, GIMP, MyPaint, and others
> pal - Aseprite and others
> col - Aseprite and others
> act - Aseprite and others
> hex - Aseprite and others
> cls (limited support) - Clip Studio Paint
> Create, modify, save, and load a variety of palette formats, allowing you to convert between them. Bring all your Krita/GIMP palettes into OpenToonz/Tahoma! (and vice versa)
> 
> Create new palettes from scratch by adding, removing ,altering and reordering entries.
> 
> Save palettes as png images to import into software such as Photoshop or Affinity Photo. Choose to display borders, border color, and the number of columns.
> 
> Extract palettes from images (Experimental). Works best with images that have a small number of colors and with no lossy compression. You can load images from local storage or from the web.
> 
> Generate new palettes by blending two colors, from random RGB ranges, and by sampling an image by clicking.
> 
> Note:
> cls support is currently limited and is planned to be improved/expanded in the future.

Note the blending of two colors line item.  This representes a stepping stone on the way to your requested feature.
Given support I imagine the Wolf that is in a bowl might be able to extend capabilities even farther.
The utility is created via Godot so there is no direct correlation between that code and Opentoonz but one never knows how the future might unfold.

There are certainly limitation in the utility.  
For instance, don't expect it to be able to manage mypaint and specialized brushes.

**Comment by RodneyBaker** [comment]
> By the way, it's technically possible paste selections in different styles? I asked this time ago, but It not was clear for me if this is not possible to Opentoonz, or is a decision.

I don't know of any way to paste into a different style but selecting strokes and applying a new style is trivial; select the strokes and then the style.

**Comment by Bracket-H** [comment]
I have always wanted a way to rearrange color indices, the only way I know is deleting and re-adding them to fill gaps and move them around.
Does this exist? 
(After the customization of room menu bars (accidentally) not enabled for Linux...I'm now going to ask more things XD)

If it doesn't ...then that would be a biiiig help (to me...) ...and if it's part of  a proper merging process, then...maybe it's a big help all around ^^

**Comment by DarrenTAnims** [comment]
You can hold the control key and then drag the palette entries to a different order or to a different tab, @Bracket-H .

**Comment by Bracket-H** [comment]
Aye, but that only changes the order of the palette, not the style indices. 

Consider the following.
![Screenshot_20220327_133041](https://user-images.githubusercontent.com/93729441/160279475-ee2858a0-a082-48bc-8a2e-bd3381c9f2b8.png)
This (crappy) drawing has indices that follow a nice, complete serial format, all in order.

Then this version:
![Screenshot_20220327_133241](https://user-images.githubusercontent.com/93729441/160279507-7d6302e0-de86-4565-a0eb-853385fbaf21.png)
Styles 3 and 4 are missing now, but on the image, they still get referenced.
How do you change Style 4 now? 
You can't unless you fill in Style 3, and then style 4, and then edit that style, because there is no way to change style indices in the image. (to my knowledge)
What if I want both missing styles to be style 2 now? 
I'd have to use the paintbrush tool to tint the lines, which is ok in such a simple drawing, but turns a hassle in more complex ones.

However!
Opentoonz already has a related option: "Delete Lines"
Where one can pick a style index, and OT will delete all pixels/lines that match the specified index.

Now, instead of deleting, what if one could give it a new index? So...
"Swap Index"
4 to 2
And now OT instead of deleting all pixels matching the index 4, it simply paints them as 2. Which then picks up the already existing palette entry.

Which, again, could also prepare levels for a merge easier.

**Comment by F-Burning** [comment]
Thank you @RodneyBaker, looks useful. However (if i understood you right) my idea is not to make a mix (red + yellow = orange), but to merge 2 equal styles into one to have all the lines in a single color.

I was trying to change the de index style too yesterday @Bracket-H, but in Toonz raster there is no way.

I always liked how RETAS PRO HD works with different colors, because you can copy them and paste into another, or simply changes it with a click.

https://youtu.be/GcPNuwEHSd4?t=149

I miss that versatility when drawing.

**Comment by Bracket-H** [comment]
@F-Burning Yeah, it's kind of a hassle...but...I just modified the underlying function that does the deleting...and 
````
void TRop::eraseColors(TRasterCM32P ras, std::vector<int> *colorIds,
                       bool eraseInks, bool noGap) {
  // noGap option is available only when erasing Inks
  assert(eraseInks || !noGap);
  if (colorIds) std::sort(colorIds->begin(), colorIds->end());

  bool hasGapPixel = false;

  for (int y = 0; y < ras->getLy(); y++) {
    TPixelCM32 *pix = ras->pixels(y), *endPix = pix + ras->getLx();

    for (; pix < endPix; pix++) {
      unsigned int i = 0;
      int color      = eraseInks ? pix->getInk() : pix->getPaint();

      if (color == 0) continue;

      if (colorIds &&
          !std::binary_search(colorIds->begin(), colorIds->end(), color))
        continue;

      if (eraseInks) {
        // no-gap case.
        // even if some area is filled under the solid line, delete it and
        // expand-fill again to make sure the colors will be separated at the
        // center of the lines.
        if (noGap && pix->getTone() == 0) {
          pix->setPaint(
              TPixelCM32::getMaxPaint());  // use paintId = 4095 as mark of
                                           // pixels to be filled
          hasGapPixel = true;
        }
        pix->setInk(0); // <========= This value sets the index of the line
        pix->setTone(TPixelCM32::getMaxTone());
      } else
        pix->setPaint(0);
    }
  }

  if (hasGapPixel) expandAreaAndFillGaps(ras);
}
````
Changing that 0 to the desired index will, at least apparently, change indices.

This means that this function could easily be extended with a new parameter, let's say "indexToSetTo", and the old functions simply call this new version of the function with a 0.
While the index changer would call it with whichever index the user wants.

**Comment by Bracket-H** [comment]
I hacked together a solution for swapping line color indices, but paint is more complicated.
Already quite the useful addition though.

**Comment by Bracket-H** [comment]
Now paint replacement also works, albeit only for replacement, deletion is buggy because it seem that there need to be some checks regarding the interplay of ink and paint.
But paint replacement seems to be perfect, it works as if one has done it diligently with the paintbrush (tint) tool.
The weird aliasing artefacts are only for deletion, the replacement workkssssss.

Also yay I didn't know I could upload videos directly.

https://user-images.githubusercontent.com/93729441/163837968-587ee51a-2068-4fe1-8380-74d85326575c.mp4

**Comment by RodneyBaker** [comment]
I've tagged this with roadmap primarily because there are some processes that merged similar colors but others that do not.  

A good next step forward will be to survey which processes require merging of similar color but currently do not.  

Also, while a defnite outlier, how can we allow for cases where the user does not want to merge similar colors when using these processes?

Adding details concerning what users are currently using as workarounds in production will be appreciated as that will help in better understanding the best way forward.

(NOTE:  This issue is ideal for Discussion so will likely be moved there)
